h2o-k8s/src/main/java/water/k8s/KubernetesEmbeddedConfigProvider.java

```diff
@@ -0,0 +1,94 @@
+
+/**
+ * A configuration provider for H2O running in Kubernetes cluster. It is able to detected H2O is being ran in K8S
+ * environment, otherwise remains inactive.
+ * <p>
+ * Uses potentially multiple strategies to discover H2O Pods on a Kubernetes cluster.
+ */
+public class KubernetesEmbeddedConfigProvider implements EmbeddedConfigProvider {
+
+    private static final String K8S_NODE_LOOKUP_TIMEOUT_KEY = "H2O_NODE_LOOKUP_TIMEOUT";
+    private static final String K8S_DESIRED_CLUSTER_SIZE_KEY = "H2O_NODE_EXPECTED_COUNT";
+
+    private boolean runningOnKubernetes = false;
+    private KubernetesEmbeddedConfig kubernetesEmbeddedConfig;
+
+    /**
+     * @return A Set
+     */
+    private static final Optional<Set<String>> resolveNodeIPs() {
+        final LookupConstraintsBuilder lookupConstraintsBuilder = new LookupConstraintsBuilder();
+
+        try {
+            final Integer timeoutSeconds = Integer.parseInt(System.getenv(K8S_NODE_LOOKUP_TIMEOUT_KEY));
+            lookupConstraintsBuilder.withTimeoutSeconds(timeoutSeconds);
+        } catch (NumberFormatException e) {
+            Log.info(String.format("'%s' environment variable not set.", K8S_NODE_LOOKUP_TIMEOUT_KEY));
+        }
+
+        try {
+            final Integer desiredClusterSize = Integer.parseInt(System.getenv(K8S_DESIRED_CLUSTER_SIZE_KEY));
+            lookupConstraintsBuilder.withDesiredClusterSize(desiredClusterSize);
+        } catch (NumberFormatException e) {
+            Log.info(String.format("'%s' environment variable not set.", K8S_DESIRED_CLUSTER_SIZE_KEY));
+        }
+
+        final KubernetesLookup kubernetesDnsDiscovery = KubernetesDnsLookup.fromH2ODefaults();
+        return kubernetesDnsDiscovery.lookupNodes(lookupConstraintsBuilder.build());
+    }
+
+    @Override
+    public void init() {
+        runningOnKubernetes = isRunningOnKubernetes();
+
+        if (!runningOnKubernetes) {
+            return; // Do not initialize any configuration if H2O is not running in K8S-spawned container.
+        }
+
+        Log.info("Initializing H2O Kubernetes cluster");
+        final Collection<String> nodeIPs = resolveNodeIPs()
+                .orElseThrow(() -> new IllegalStateException("Unable to resolve Node IPs from DNS service."));
+
+        Log.info(String.format("Using the following pods to form H2O cluster: [%s]",
+                String.join(",", nodeIPs)));
+
+        kubernetesEmbeddedConfig = new KubernetesEmbeddedConfig(nodeIPs);
+    }
+
+    @Override
+    public boolean isActive() {
+        return runningOnKubernetes;
+    }
+
+    @Override
+    public AbstractEmbeddedH2OConfig getConfig() {
+        return kubernetesEmbeddedConfig;
+    }
+
+    /**
+     * @return True if there are environment variables indicating H2O is running inside a container managed by
+     * Kubernetes. Otherwise false.
+     */
+    private boolean isRunningOnKubernetes() {
+        final Pattern KUBERNETES_SERVICE_HOST = Pattern.compile(".*_SERVICE_HOST");
```

Comment:
QQ: are these guaranteed to exist in every k8s environment?

Reply:
According to documentation, the answer is yes: https://kubernetes.io/docs/concepts/services-networking/service/#discovering-services

Reply:
This is very K8S-specific and looks like a good indicator that H2O is running on K8S cluster, unless a better one is found.